x-pack/platform/plugins/shared/stack_alerts/common/es_query/esql_query_utils.ts

```diff
@@ -63,37 +64,61 @@ export const ALERT_ID_COLUMN = 'Alert ID';
 export const ALERT_ID_SUGGESTED_MAX = 10;
 
 export const rowToDocument = (columns: EsqlResultColumn[], row: EsqlResultRow): EsqlDocument => {
-  return columns.reduce<Record<string, string | null>>((acc, column, i) => {
-    acc[column.name] = row[i];
-
-    return acc;
-  }, {});
+  const doc: EsqlDocument = {};
+  for (let i = 0; i < columns.length; ++i) {
+    doc[columns[i].name] = row[i];
+  }
+  return doc;
 };
 
-export const getEsqlQueryHits = (
+export const getEsqlQueryHits = async (
   table: EsqlTable,
   query: string,
-  isGroupAgg: boolean
-): EsqlQueryHits => {
+  isGroupAgg: boolean,
+  isPreview: boolean = false
+): Promise<EsqlQueryHits> => {
   if (isGroupAgg) {
     const alertIdFields = getAlertIdFields(query, table.columns);
-    return toGroupedEsqlQueryHits(table, alertIdFields);
+    return toGroupedEsqlQueryHits(table, alertIdFields, isPreview);
   }
-  return toEsqlQueryHits(table);
+  return toEsqlQueryHits(table, isPreview);
 };
 
-export const toEsqlQueryHits = (table: EsqlTable): EsqlQueryHits => {
+export const toEsqlQueryHits = async (
+  table: EsqlTable,
+  isPreview: boolean = false,
+  chunkSize: number = CHUNK_SIZE
+): Promise<EsqlQueryHits> => {
   const hits: EsqlHit[] = [];
   const rows: EsqlDocument[] = [];
-  for (const row of table.values) {
-    const document = rowToDocument(table.columns, row);
-    hits.push({
-      _id: ESQL_DOCUMENT_ID,
-      _index: '',
-      _source: document,
-    });
-    rows.push(rows.length > 0 ? document : { [ALERT_ID_COLUMN]: ActionGroupId, ...document });
-  }
+  const numRows = table.values.length;
+  let currentRow = 0;
+
+  const processEsqlQueryHits = async () => {
+    const chunk = Math.min(currentRow + chunkSize, numRows);
+    for (; currentRow < chunk; currentRow++) {
+      const row = table.values[currentRow];
+      const document = rowToDocument(table.columns, row);
+      hits.push({
+        _id: ESQL_DOCUMENT_ID,
+        _index: '',
+        _source: document,
+      });
+      if (isPreview) {
+        rows.push(
+          rows.length > 0 ? document : Object.assign({ [ALERT_ID_COLUMN]: ActionGroupId }, document)
+        );
+      }
+    }
+
+    if (currentRow < numRows) {
+      await new Promise((resolve) => {
+        setImmediate(() => resolve(processEsqlQueryHits()));
```

Comment:
I wrapped this in a promise, but pls let me know if that's what you would do here? I looked in Kibana and it seems like this is how other areas implemented `setImmediate`